tracer/src/Datadog.Trace/TraceProcessors/TraceUtil.cs

```diff
@@ -0,0 +1,286 @@
+                }
+                else
+                {
+                    if (cuts is null)
+                    {
+                        cuts = new List<int[]>();
+                    }
+
+                    var n = cuts.Count;
+                    if (n > 0 && cuts[n - 1][1] >= i)
+                    {
+                        cuts[n - 1][1] += jump;
+                    }
+                    else
+                    {
+                        cuts.Add(new int[] { i, i + jump });
+                    }
+                }
+
+            end:
+                if (i + jump >= 2 * MaxTagLength)
+                {
+                    i++;
+                    break;
+                }
+
+                if (chars >= MaxTagLength)
+                {
+                    i++;
+                    break;
+                }
+            }
+
+            i--;
+
+            if (cuts is null || cuts.Count == 0)
+            {
+                if (charArray is null)
+                {
+                    return value.Substring(trim, i + jump - trim);
+                }
+
+                return new string(charArray, trim, i + jump - trim);
+            }
+
+            charArray ??= value.ToCharArray();
+            var segment = new ArraySlice<char>(charArray, trim, i + jump - trim);
+            int delta = trim;
+            foreach (var cut in cuts)
+            {
+                int start = cut[0] - delta;
+                int end = cut[1] - delta;
+
+                if (end >= segment.Count)
+                {
+                    segment = segment.Slice(0, start);
+                    break;
+                }
+
+                segment[start] = '_';
+                if (end - start == 1)
+                {
+                    continue;
+                }
+
+                segment.Slice(end).CopyTo(segment.Slice(start + 1));
+                segment = segment.Slice(0, segment.Count - (end - start) + 1);
+                delta += cut[1] - cut[0] - 1;
+            }
+
+            return new string(segment.Array, segment.Offset, segment.Count);
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/traceutil/normalize.go#L213-L216
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsAlpha(char c)
+        {
+            return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z');
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/traceutil/normalize.go#L218-L221
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsAlphaNum(char c)
+        {
+            return IsAlpha(c) || (c >= '0' && c <= '9');
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/traceutil/normalize.go#L223-L274
+        public static string NormMetricNameParse(string name, int limit)
```

Comment:
nit: could we expand this name a bit? `NormalizeMetricName`?

Reply:
done in [b4d13f1](https://github.com/DataDog/dd-trace-dotnet/pull/2402/commits/b4d13f132f14eaddc035abfb75ae6631efac94f2)